lib/likelihood.c

```diff
@@ -0,0 +1,199 @@
+/*
+** Copyright (C) 2019 University of Oxford
+**
+** This file is part of msprime.
+**
+** msprime is free software: you can redistribute it and/or modify
+** it under the terms of the GNU General Public License as published by
+** the Free Software Foundation, either version 3 of the License, or
+** (at your option) any later version.
+**
+** msprime is distributed in the hope that it will be useful,
+** but WITHOUT ANY WARRANTY; without even the implied warranty of
+** MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+** GNU General Public License for more details.
+**
+** You should have received a copy of the GNU General Public License
+** along with msprime.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+/*
+ * Implementation of log-likelihood functions for full-ARG.
+ */
+
+#include <float.h>
+#include <math.h>
+#include <stdlib.h>
+#include "likelihood.h"
+#include "msprime.h"
+
+static double
+get_total_material(tsk_treeseq_t *ts) {
+    double ret = 0;
+    tsk_size_t j;
+    tsk_edge_table_t *edges = &ts->tables->edges;
+    tsk_node_table_t *nodes = &ts->tables->nodes;
+    tsk_id_t parent;
+    tsk_id_t child;
+    
+    for (j = 0; j < edges->num_rows; j++) {
+        parent = edges->parent[j];
+        child = edges->child[j];
+        ret += (edges->right[j] - edges->left[j])
+            * (nodes->time[parent] - nodes->time[child]);
+    }
+    return ret;
+}
+
+int
+msp_unnormalised_log_likelihood_mut(tsk_treeseq_t *ts, double theta,
+                                    double *r_lik) {
+    int ret = 1;
+    tsk_id_t it;
+    tsk_size_t j;
+    tsk_mutation_t mut;
+    tsk_size_t num_mutations = tsk_treeseq_get_num_mutations(ts);
+    const double total_material = get_total_material(ts);
+    const double *node_time = ts->tables->nodes.time;
+    double branch_length;
```

Comment:
can declare branch_length and lik on the same line here, they're simple local variables that don't need initialialisation